webapp/app.py

```diff
@@ -23,94 +27,129 @@
     category_id=15,  # The "doc" category
 )
 
-app = flask.Flask(__name__)
-app.template_folder = "../templates"
-app.static_folder = "../static"
-app.url_map.strict_slashes = False
 
-# Parse redirects.yaml and permanent-redirects.yaml
-app.before_request(prepare_redirects())
+def create_app(testing=False):
```

Comment:
I haven't got around to discussing this pattern with @tbille - personally I don't understand why it's needed, and I think it's inferior to simply declaring the `app` directly.

Until we've discussed it, I'd like to not consider this structure as a standard and so limit its use to snapcraft.io. Could you please remove it again from this project for now?

Reply:
This was the only way I had to make tests work properly both locally and in the CI because it allows me to set app.testing and app.debug early on. I feel like if you don't you get some unexpected behavior, particularly exception propagation and things like that.

Reply:
`app.debug` is set with `FLASK_DEBUG=true` - this should have exactly the same effect, and if you need debug mode you should set it with this environment variable. But I don't see why tests should need `app.debug` - tests should surely run under more production-like settings shouldn't they?

`app.testing` is described [in the codebase](https://github.com/pallets/flask/blob/master/flask/app.py#L955):

> Note that if you are testing for assertions or exceptions in your application code, you must set ``app.testing = True`` in order for the exceptions to propagate to the test client.  Otherwise, the exception will be handled by the application (not visible to the test client) and the only indication of an AssertionError or other exception will be a 500 status code response to the test client.  See the :attr:`testing` attribute.

Why can't this be set inside the tests themselves, as I did [in blog.ubuntu.com](https://github.com/canonical-websites/blog.ubuntu.com/blob/master/tests/tests.py#L77)

``` python3
    def setUp(self):
        self.app = app.app.test_client()
        self.app.testing = True
```

If this doesn't work, could you please demonstrate it to me?

My objection to doing it the `create_app` way is I think it's very undesirable to complexify your actual app with things that are just needed for testing. Of course tests sometimes lead us to change our app by e.g. improving dependency injection, but that should only be done if it's a fundamentally better structure for the application. Plus the `create_app` pattern is not particularly common, directly creating your WSGI app is common both throughout Flask projects and throughout WSGI projects.

Reply:
I'm not running tests with debug on, I'm forcing it off regardless of what `FLASK_DEBUG` says.

If you checkout my branch and delete line 35 of `app.py` and `./run test` the tests will fail.
They will keep failing even if you set `app.debug = False` in the test setUp because the app was initialized with `debug = True` picked from `.env.local`.

Also I don't think the pattern is that uncommon, it's actually used in the docs(http://flask.pocoo.org/docs/1.0/tutorial/factory/) and I think we could actually progress to storing each environment config in a file and just load it in this function.

I don't think it adds much complexity to be honest, just wraps the current logic in a function.

Reply:
> I'm not running tests with debug on, I'm forcing it off regardless of what FLASK_DEBUG says.

Ah fair enough, sorry I misinterpreted that. That's a good point, it's probably a good idea to run tests without debug mode. And I can see that debug mode causes all sort of problems - it doesn't return errors properly etc. And you're right, setting `app.debug` in the test itself appears to be too late.

But this could also be fixed through changing `package.json` to:

``` json
"test-python": "FLASK_DEBUG=false coverage run --source=. -m unittest discover webapp/tests",
```

I don't think it's unreasonable to push the responsibility to run the tests with the right environment settings back to the client. After all, they could get very confused about why they're adding `FLASK_DEBUG=true` but then experiencing `app.debug` being false in the app. We could, if you wanted, enforce this by adding a check to the test:

``` python3
if app.debug == True:
    raise Exception("Tests must be run with debug mode turned off")
```

> I don't think it adds much complexity to be honest, just wraps the current logic in a function.

I agree it's not *very* complex, but even the slightest complexity has a cost to it and so we should never be adding more complexity than we need. Especially when it's a baseline pattern that we intend to use in every Flask app.

I'm aware this pattern is offered in http://flask.pocoo.org/docs/1.0/tutorial/factory/ as an option if you need it, but that doc is far from recommending it as a standard. And I haven't seen it in the wild all that much.

I'm afraid I still feel very strongly that testing logic doesn't belong in the application if it's at all avoidable - tests should bend over backwards to remain external to the application - and in this case it appears to me that it's avoidable.

Reply:
So setting `FLASK_DEBUG=false` in the package.json does not work because when you don't use the factory the same app instance is shared by all tests, which means if the tests that set the search key in the app config run first than the ones that need a search key not to be set we will have errors.

Reply:
@nottrobin Reverted the change and made the tests work. Going to ask @tbille to approve if you didn't have any concerns apart from this one. Is that alright?